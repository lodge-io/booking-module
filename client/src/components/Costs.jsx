import React2 from 'react';
import styled2 from 'styled-components';

const React = window.React || React2;
const styled = window.styled || styled2;

const Row = styled.div`
  display: flex;
`;

const RowItem = styled.div`
  flex-grow:1;
  text-align: ${props => (props.isCost ? 'right' : 'left')}
`;

const CostLine = ({ title, cost, isTotal }) => (
  <div>
    <Row>
      <RowItem isCost={false}>
        {title}
      </RowItem>
      <RowItem className={isTotal ? 'totalBookingCost' : ''} isCost>
        {`$${cost}`}
      </RowItem>
    </Row>
    {isTotal ? '' : <hr />}
  </div>
);

function getTaxRate(taxes) {
  let rate = 0;
  taxes.forEach((tax) => { rate += tax.type === 'percent' ? tax.rate : 0; });
  return rate;
}

function getFlatTax(taxes) {
  let total = 0;

  taxes.forEach((tax) => { total += tax.type === 'flat' ? tax.amount : 0; });

  return total;
}


const Costs = ({
  duration, price, fees, taxes 
}) => {
  const baseStayCost = price * duration;
  const feeArr = [];
  let totalFeeCost = 0;
  Object.keys(fees).forEach((fee) => { feeArr.push([fee, fees[fee]]); totalFeeCost += fees[fee]; });

  const totalTaxCost = parseInt(((baseStayCost + totalFeeCost)
    * getTaxRate(taxes)
    + getFlatTax(taxes)).toFixed(), 10);

  const totalCost = baseStayCost + totalFeeCost + totalTaxCost;
  return (
    <div>
      <CostLine title={`$${price} x ${duration} nights`} cost={baseStayCost} />
      {feeArr.map(fee => (
        <CostLine title={fee[0]} cost={fee[1]} />
      ))}
      <CostLine title="Occupancy Taxes and Fees" cost={totalTaxCost} />
      <CostLine title="Total" cost={totalCost} isTotal />
    </div>
  );
};

export default Costs;
