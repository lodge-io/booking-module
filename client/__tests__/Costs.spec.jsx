import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Costs from '../src/components/Costs';

describe('Price', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Costs duration={1} price={20} fees={{ 'Fee 1': 20 }} taxes={[{ name: 'Evil Tax', type: 'flat', amount: 1000 }]} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should show total based on price and duration', () => {
    const wrapper = mount(
      <Costs
        duration={1}
        price={20}
        fees={{ 'Fee 1': 20, 'Clening Fee': 30, 'Service Fee': 20 }}
        taxes={[{ name: 'Evil Tax', type: 'flat', amount: 1000 }, { name: 'Good Tax', type:'percent', amount: '0.12' }]} 
      />,
    );
    const calc = (20 * 1 + 20 + 30 + 20) * 1.12 + 1000;
    // expect(wrapper.find('.totalBookingCost').contains(`$${calc}`)).toBeTruthy();
    wrapper.unmount();
  });
});
