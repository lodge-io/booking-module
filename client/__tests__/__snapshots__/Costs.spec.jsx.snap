// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price renders correctly 1`] = `
<div>
  <div>
    <div
      className="sc-bdVaJa kZAgjJ"
    >
      <div
        className="sc-bwzfXH gOZtLQ"
      >
        $20 x 1 nights
      </div>
      <div
        className="sc-bwzfXH bQEDAm"
      >
        $20
      </div>
    </div>
    <hr />
  </div>
  <div>
    <div
      className="sc-bdVaJa kZAgjJ"
    >
      <div
        className="sc-bwzfXH gOZtLQ"
      >
        Fee 1
      </div>
      <div
        className="sc-bwzfXH bQEDAm"
      >
        $20
      </div>
    </div>
    <hr />
  </div>
  <div>
    <div
      className="sc-bdVaJa kZAgjJ"
    >
      <div
        className="sc-bwzfXH gOZtLQ"
      >
        Occupancy Taxes and Fees
      </div>
      <div
        className="sc-bwzfXH bQEDAm"
      >
        $1000
      </div>
    </div>
    <hr />
  </div>
  <div>
    <div
      className="sc-bdVaJa kZAgjJ"
    >
      <div
        className="sc-bwzfXH gOZtLQ"
      >
        Total
      </div>
      <div
        className="totalBookingCost sc-bwzfXH bQEDAm"
      >
        $1040
      </div>
    </div>
    
  </div>
</div>
`;
